search/src/search.rs

```diff
@@ -164,10 +169,103 @@ impl Search {
+            .index
+            .searcher()
+            .map_err(|e| SearchError(e.to_string()))?;
+        let query = self
+            .index
+            .query_parser()
+            .map_err(|e| SearchError(e.to_string()))?
+            .parse_query(term)
+            .map_err(|e| SearchError(e.to_string()))?;
+        let mut snippet_generator = SnippetGenerator::create(&searcher, &*query, self.schema.body)
+            .map_err(|e| SearchError(e.to_string()))?;
+        snippet_generator.set_max_num_chars(70);
+
         let snippet = snippet_generator.snippet(&text);
         // let snippet = snippet_generator.snippet_from_doc(&retrieved_doc);
-        Ok(self.highlight(snippet))
+        Ok((self.highlight(snippet), get_first_term(term)))
+    }
+
+    fn get_fuzzy_preview(
+        &self,
+        path: &str,
+        term: &str,
+        distance: &u8,
+        body: &str,
+    ) -> Result<(String, String)> {
+        let pindex: PageIndex = self
+            .index
+            .page_index(path)
+            .map_err(|_| SearchError("".to_string()))?;
+        let (matched_term, start, end) = self
+            .get_fuzzy_match(term, distance, pindex)
+            .map_err(|_| SearchError("".to_string()))?;
+        // Another safe way to get substrings using char_indices
+        let start = body
+            .char_indices()
+            .nth(start.saturating_sub(20) as usize)
+            .unwrap_or((0, ' '))
+            .0;
+        let end = body
+            .char_indices()
+            .nth((end + 20) as usize)
+            .unwrap_or((body.len() - 1, ' '))
+            .0;
+        let substring = &format!("...{}...", &body[start..end]);
+        let substring = substring
+            .to_string()
+            .replace(&matched_term, &format!("**{}**", matched_term));
+        Ok((substring.replace('\n', " "), matched_term))
+    }
+
+    fn get_fuzzy_match(
+        &self,
+        term: &str,
+        distance: &u8,
+        pindex: PageIndex,
+    ) -> Result<(String, u32, u32)> {
+        if pindex.contains_key(term) {
+            let (start, end) = pindex.get(term).unwrap().first().unwrap();
+            Ok((term.to_string(), *start, *end))
+        } else {
+            let mut cur: (String, u32, u32) = ("".to_string(), 0, 0);
+            let mut min_dist: usize = usize::MAX;
+            for (word, matches) in pindex {
+                let dist: usize = levenshtein(term, &word);
+                let dist = if word.contains(term) { 1 } else { dist };
+                println!("{} ~ {} = {}", term, word, dist);
+                if dist < min_dist {
+                    min_dist = dist;
+                    let (start, end) = matches.first().unwrap_or(&(0, 0));
+                    cur = (word.to_string(), *start, *end)
+                }
+            }
+            if min_dist as u8 <= *distance {
+                Ok(cur)
+            } else {
+                Err(SearchError("".to_string()))
+            }
```

Comment:
I think I'm missing something here, can you help me understand?
To me it looks like this should ensure that we're not getting matches with a distance >= distance. However, in the tests, we do. Why?

Reply:
Yes, the first part is correct: After finding the word with the minimal distance to the searched term, we ensure it is within the range of the maximum distance defined by the user (default: `2`).  If not, we throw an error which is handled `get_preview` function, returning `FUZZY_PREVIEW_NOT_FOUND` as preview. 

However, I don't see, where you see greater distances in the tests?